rootfs/scheduler/__init__.py

```diff
@@ -568,42 +583,53 @@ def run(self, namespace, name, image, entrypoint, command, **kwargs):
 
-        data = ''
-        duration = 30
-        iteration = 1
-        while (iteration < duration):
-            try:
+        labels = {
+            'app': namespace,
+            'type': 'run',
+            'version': kwargs.get('version'),
+            'heritage': 'deis',
+        }
+        # wait for run pod to start - use the same function as scale
+        self._wait_until_pods_are_ready(namespace, container, labels, desired=1)
+
+        try:
+            # give pod 20 minutes to execute (after it got into ready state)
+            # this is a fairly arbitrary limit but the gunicorn worker / LBs
+            # will make this timeout around 20 anyway.
+            # TODO: Revisit in the future so it can run longer
+            state = 'up'  # pod is still running
+            waited = 0
+            timeout = 1200  # 20 minutes
+            while (state == 'up' and waited < timeout):
                 response = self._get_pod(namespace, name)
-                data = response.text
                 pod = response.json()
-                if pod['status']['phase'] == 'Succeeded':
-                    response = self._pod_log(namespace, name)
-                    response.encoding = 'utf-8'  # defaults to "ISO-8859-1" otherwise...
-                    log = response.text
-                    self._delete_pod(namespace, name)
-                    return 0, log
-
-                elif pod['status']['phase'] == 'Running':
-                    if iteration > 28:
-                        duration += 1
-
-                elif pod['status']['phase'] == 'Failed':
-                    pod_state = pod['status']['containerStatuses'][0]['state']
-                    err_code = pod_state['terminated']['exitCode']
-                    self._delete_pod(namespace, name)
-                    return err_code, data
+                state = self._pod_state(pod).name
+                # default data
+                exit_code = 0
 
-            except KubeException:
-                break
+                waited += 1
+                time.sleep(1)
 
-            iteration += 1
-            time.sleep(1)
+            if state == 'down':  # run finished successfully
+                exit_code = 0  # successful run
+            elif state == 'crashed':  # run failed
+                pod_state = pod['status']['containerStatuses'][0]['state']
+                exit_code = pod_state['terminated']['exitCode']
 
-        if iteration >= duration:
-            error(response, 'Pod start took more than 30 seconds', namespace)
-            return 0, data
+            # timed out!
+            if waited == timeout:
+                raise KubeException('Timed out (20 mins) while running')
 
-        return 0, data
+            # grab log information
+            log = self._pod_log(namespace, name)
+            log.encoding = 'utf-8'  # defaults to "ISO-8859-1" otherwise...
+
+            return exit_code, log.text
+        except KubeException:
```

Comment:
Does that play with `finally`? I can give it a go just to see, usually have it coupled with `except`